web/js/codeworld_shared.js

```diff
@@ -225,31 +291,58 @@ function registerStandardHints(successFunc)
         }
 
         var found = [];
-
+        var hints = Object.keys(codeWorldSymbols)
         for (var i = 0; i < hints.length; i++) {
             var hint = hints[i];
-            if (hint.text.startsWith(term)) {
+            if (hint.startsWith(term)){
                 found.push(hint);
             }
         }
 
-        var lines = cm.getValue().split("\n");
-        for (var i=0; i < lines.length; i++) {
-            if (/^\S*\s*::[^:]*$/.test(lines[i])) {
-                var candidate = lines[i].split(" ::")[0];
-                if (candidate.startsWith(term)) {
-                    found.push(createHint(lines[i], 0, candidate.length));
-                }
+        found.sort(function(a, b) {
+            function startsWithLetter(c) {
+                return /^[a-zA-Z].*/.test(c);
             }
-        }
+                if (startsWithLetter(a) && !startsWithLetter(b)) return -1;
+                else if (startsWithLetter(b) && !startsWithLetter(a)) return 1;
+                else return a.toLowerCase() < b.toLowerCase() ? -1 : 1
+            });
 
-        sortHints(found);
+        if (found.length > 0) {
+            var data = {
+                list: found,
+                from: from,
+                to: cur
+            };
 
-        if (found.length > 0) return {
-            list: found,
-            from: from,
-            to: cur
-        };
+            function deleteOldHintDocs(){
+                $(".hint-description").remove()
+            }
+
+            CodeMirror.on(data, 'close', deleteOldHintDocs);
+            CodeMirror.on(data, 'pick', deleteOldHintDocs);
+
+            // Tracking of hint selection
+            CodeMirror.on(
+                data, 'select',
```

Comment:
I can't find any clear documentation on this form of `on`.  Everything I'm looking at uses the first parameter to say which CodeMirror instance you're adding the listener to.  What does it mean to pass `data` there?

Reply:
`'select'`  is event from [hint/show-hint.js addon](https://codemirror.net/doc/manual.html#addon_show-hint).
"The following events will be fired on the completions object during completion: " 
Data here is "completions object", it is not obvious from documentation. I understood how it works after reading https://github.com/codemirror/CodeMirror/issues/3092 .